### PR TITLE
Adapt TabBar to new DPI management

### PR DESCRIPTION
**Pull request by donho**
ref #14959

### CODE REVIEW

**Comment by donho** [comment]
@ozone10 
It doesn't work even scale on the same monitor.
I do believe it's due to the code of `case WM_DPICHANGED:`
Could you provide some suggestion?

File: PowerEditor/src/Notepad_plus.cpp [hunk 1]
```diff
@@ -382,25 +383,25 @@ LRESULT Notepad_plus::init(HWND hwnd)
...
 	if (hf)
 	{
 		::SendMessage(_mainDocTab.getHSelf(), WM_SETFONT, reinterpret_cast<WPARAM>(hf), MAKELPARAM(TRUE, 0));
 		::SendMessage(_subDocTab.getHSelf(), WM_SETFONT, reinterpret_cast<WPARAM>(hf), MAKELPARAM(TRUE, 0));
 	}
 
-	int tabDpiDynamicalHeight = nppParam._dpiManager.scaleY(_toReduceTabBar ? g_TabHeight : g_TabHeightLarge);
-	int tabDpiDynamicalWidth = nppParam._dpiManager.scaleX(g_TabWidth);
+	int tabDpiDynamicalHeight = _mainDocTab.dpiManager().scale(TabBarPlus::isReduced() ? g_TabHeight : g_TabHeightLarge);
+	int tabDpiDynamicalWidth = _mainDocTab.dpiManager().scale(g_TabWidth);
```
Comment (ozone10) [inline PowerEditor/src/Notepad_plus.cpp:394]: missing close button width check
```c++
int tabDpiDynamicalWidth = _mainDocTab.dpiManager().scale(TabBarPlus::drawTabCloseButton() ? g_TabWidthCloseBtn : g_TabWidth);
```
and should also handle padding
```c++
TabCtrl_SetPadding(_mainDocTab.getHSelf(), _mainDocTab.dpiManager().scale(TabBarPlus::drawTabCloseButton() ? 10 : 6), 0);
```
values from `void TabBarPlus::doOwnerDrawTab(TabBarPlus* tbpObj)`

File: PowerEditor/src/NppCommands.cpp [hunk 2]
```diff
@@ -2273,17 +2273,16 @@ void Notepad_plus::command(int id)
 
 		case IDM_VIEW_REDUCETABBAR:
```
Comment (ozone10) [inline PowerEditor/src/NppCommands.cpp:2274]: should also handle padding
```c++
TabCtrl_SetPadding(_mainDocTab.getHSelf(), _mainDocTab.dpiManager().scale(TabBarPlus::drawTabCloseButton() ? 10 : 6), 0);
```
values from `void TabBarPlus::doOwnerDrawTab(TabBarPlus* tbpObj)`

File: PowerEditor/src/NppCommands.cpp [hunk 3]
```diff
@@ -2273,17 +2273,16 @@ void Notepad_plus::command(int id)
...
 		{
-			_toReduceTabBar = !_toReduceTabBar;
-			auto& dpiManager = NppParameters::getInstance()._dpiManager;
+			bool isReduceed = TabBarPlus::isReduced();
 
 			//Resize the tab height
-			int tabDpiDynamicalWidth = dpiManager.scaleX(g_TabWidth);
-			int tabDpiDynamicalHeight = dpiManager.scaleY(_toReduceTabBar ? g_TabHeight : g_TabHeightLarge);
+			int tabDpiDynamicalWidth = _mainDocTab.dpiManager().scale(g_TabWidth);
```
Comment (ozone10) [inline PowerEditor/src/NppCommands.cpp:2279]: missing close button width check
```c++
int tabDpiDynamicalWidth = _mainDocTab.dpiManager().scale(TabBarPlus::drawTabCloseButton() ? g_TabWidthCloseBtn : g_TabWidth);
```

File: PowerEditor/src/WinControls/TabBar/TabBar.cpp [hunk 4]
```diff
@@ -369,29 +339,49 @@ void TabBarPlus::init(HINSTANCE hInst, HWND parent, bool isVertical, bool isMult
...
 	lfVer.lfWeight = FW_HEAVY;
+
 	_hVerticalLargeFont = CreateFontIndirect(&lfVer);
 }
 
 
-void TabBarPlus::doOwnerDrawTab()
+void TabBarPlus::doOwnerDrawTab(TabBarPlus* tbpObj)
 {
 	::SendMessage(_hwndArray[0], TCM_SETPADDING, 0, MAKELPARAM(6, 0));
```
Comment (ozone10) [inline PowerEditor/src/WinControls/TabBar/TabBar.cpp:386]: can be removed, due to line 404

File: PowerEditor/src/NppBigSwitch.cpp [hunk 5]
```diff
@@ -3516,6 +3516,12 @@ LRESULT Notepad_plus::process(HWND hwnd, UINT message, WPARAM wParam, LPARAM lPa
 			const UINT dpi = LOWORD(wParam);
 			_toolBar.resizeIconsDpi(dpi);
 
+			_mainDocTab.dpiManager().setDpi(dpi);
+			_subDocTab.dpiManager().setDpi(dpi);
+			::SendMessage(_pPublicInterface->getHSelf(), WM_COMMAND, IDM_VIEW_REDUCETABBAR, 0);
+			_mainDocTab.setFont();
+			_subDocTab.setFont();
```
Comment (ozone10) [inline PowerEditor/src/NppBigSwitch.cpp:3523]: both `setFont()` should be before `::SendMessage(_pPublicInterface->getHSelf(), WM_COMMAND, IDM_VIEW_REDUCETABBAR, 0);`

**Comment by donho** [comment]
@ozone10 
Thank you for your psychic debugging! A lot of bugs are fixed in the latest commit.
It remains only the icon/imagelist resizing in DocTab code part.
I'll be back in a few hours for working on it.

**Comment by donho** [comment]
@ozone10 
I expected the latest commit makes tabbar icons resized correctly. However, after changing the DPI, tabbar icons disappear completely. 
Could you use your psychic debugging power to detect the problem? Thank you in advance!

File: PowerEditor/src/NppCommands.cpp [hunk 6]
```diff
@@ -2273,17 +2273,19 @@ void Notepad_plus::command(int id)
...
-			auto& dpiManager = NppParameters::getInstance()._dpiManager;
+			bool isReduceed = TabBarPlus::isReduced();
 
 			//Resize the tab height
-			int tabDpiDynamicalWidth = dpiManager.scaleX(g_TabWidth);
-			int tabDpiDynamicalHeight = dpiManager.scaleY(_toReduceTabBar ? g_TabHeight : g_TabHeightLarge);
+			int tabDpiDynamicalWidth = _mainDocTab.dpiManager().scale(TabBarPlus::drawTabCloseButton() ? g_TabWidthCloseBtn : g_TabWidth);
+			int tabDpiDynamicalHeight = _mainDocTab.dpiManager().scale(isReduceed ? g_TabHeight : g_TabHeightLarge);
+			
+			TabCtrl_SetPadding(_mainDocTab.getHSelf(), _mainDocTab.dpiManager().scale(TabBarPlus::drawTabCloseButton() ? 10 : 6), 0);
```
Comment (ozone10) [inline PowerEditor/src/NppCommands.cpp:2282]: missing padding for subview

File: PowerEditor/src/ScintillaComponent/DocTabView.h [hunk 7]
```diff
@@ -86,17 +72,37 @@ public :
 
 	void reSizeTo(RECT & rc) override;
 
+	void resizeIconsDpi() {
+		UINT newSize = dpiManager().scale(g_TabIconSize);
+		for (auto i : _pIconListVector)
```
Comment (ozone10) [inline PowerEditor/src/ScintillaComponent/DocTabView.h:77]: use `for (const auto& i : _pIconListVector)`

File: PowerEditor/src/ScintillaComponent/DocTabView.h [hunk 8]
```diff
@@ -86,17 +72,37 @@ public :
...
 	void reSizeTo(RECT & rc) override;
 
+	void resizeIconsDpi() {
+		UINT newSize = dpiManager().scale(g_TabIconSize);
+		for (auto i : _pIconListVector)
+		{
+			ImageList_SetIconSize(i->getHandle(), newSize, newSize);
+		}
+
+		reInit();
```
Comment (ozone10) [inline PowerEditor/src/ScintillaComponent/DocTabView.h:82]: add below
```c++
if (_iconListIndexChoice != -1)
	TabBar::setImageList(_pIconListVector[_iconListIndexChoice]->getHandle());
```

**Comment by ozone10** [comment]
@donho 
check
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/e877707ecb58fb9e7b4ba5a44b3bc35aa248f980/PowerEditor/src/NppBigSwitch.cpp#L918

for how to handle icons for document list 

---
also this is something I had prepared before for toolbar icons, but it should also improve icons for tabs

change 
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/e877707ecb58fb9e7b4ba5a44b3bc35aa248f980/PowerEditor/src/WinControls/ImageListSet/ImageListSet.h#L42
to `void addIcon(int iconID, int cx = 32, int cy = 32) const;`

change 
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/e877707ecb58fb9e7b4ba5a44b3bc35aa248f980/PowerEditor/src/WinControls/ImageListSet/ImageListSet.cpp#L34

https://github.com/notepad-plus-plus/notepad-plus-plus/blob/e877707ecb58fb9e7b4ba5a44b3bc35aa248f980/PowerEditor/src/WinControls/ImageListSet/ImageListSet.cpp#L44

to 
```c++
void IconList::create(int iconSize, HINSTANCE hInst, int *iconIDArray, int iconIDArraySize)
{
	init(hInst, iconSize);
	_pIconIDArray = iconIDArray;
	_iconIDArraySize = iconIDArraySize;

	for (int i = 0 ; i < iconIDArraySize ; ++i)
		addIcon(iconIDArray[i], iconSize, iconSize);
}

void IconList::addIcon(int iconID, int cx, int cy) const 
{
	HICON hIcon = nullptr;
	DPIManagerV2::loadIcon(_hInst, MAKEINTRESOURCE(iconID), cx, cy, &hIcon, LR_DEFAULTSIZE);

	if (!hIcon)
		throw std::runtime_error("IconList::addIcon : LoadIcon() function return null");

	ImageList_AddIcon(_hImglst, hIcon);
	::DestroyIcon(hIcon);
}
```

modify so it use size
https://github.com/notepad-plus-plus/notepad-plus-plus/blob/e877707ecb58fb9e7b4ba5a44b3bc35aa248f980/PowerEditor/src/WinControls/ImageListSet/ImageListSet.cpp#L102

```c++
_iconListVector[HLIST_DEFAULT].addIcon(_tbiis[i]._defaultIcon, size, size);
_iconListVector[HLIST_DISABLE].addIcon(_tbiis[i]._grayIcon, size, size);
_iconListVector[HLIST_DEFAULT2].addIcon(_tbiis[i]._defaultIcon2, size, size);
_iconListVector[HLIST_DISABLE2].addIcon(_tbiis[i]._grayIcon2, size, size);

_iconListVector[HLIST_DEFAULT_DM].addIcon(_tbiis[i]._defaultDarkModeIcon, size, size);
_iconListVector[HLIST_DISABLE_DM].addIcon(_tbiis[i]._grayDarkModeIcon, size, size);
_iconListVector[HLIST_DEFAULT_DM2].addIcon(_tbiis[i]._defaultDarkModeIcon2, size, size);
_iconListVector[HLIST_DISABLE_DM2].addIcon(_tbiis[i]._grayDarkModeIcon2, size, size);
```

File: PowerEditor/src/ScintillaComponent/DocTabView.h (hunk 8 above)
Comment (donho) [reply PowerEditor/src/ScintillaComponent/DocTabView.h]: @ozone10 
Thank you for this magic line!

**Comment by donho** [comment]
> @donho check ...

All done in the latest commit.
Thank you for the code review.

File: PowerEditor/src/resource.h [hunk 9]
```diff
@@ -609,7 +609,7 @@
 	#define NPPM_INTERNAL_RELOADNATIVELANG              (NOTEPADPLUS_USER_INTERNAL + 25)
 	#define NPPM_INTERNAL_PLUGINSHORTCUTMOTIFIED        (NOTEPADPLUS_USER_INTERNAL + 26)
 	#define NPPM_INTERNAL_SCINTILLAFINDERCLEARALL       (NOTEPADPLUS_USER_INTERNAL + 27)
-	#define NPPM_INTERNAL_CHANGETABBAEICONS             (NOTEPADPLUS_USER_INTERNAL + 28)
+	#define NPPM_INTERNAL_CHANGETABBAEICONSET           (NOTEPADPLUS_USER_INTERNAL + 28)
```
Comment (MarkusBodensee) [inline PowerEditor/src/resource.h:612]: minor typo --> NPPM_INTERNAL_CHANGETABBARICONSET

**Pull request closed without merging**